### PR TITLE
Update klog to 0.3.0

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
This PR updates `klog` dependency to the latest release: 0.3.0.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Updated klog to 0.3.0
```

### CODE REVIEW

**Review by dims — APPROVED** [review]

File: cmd/kubelet/app/options/globalflags.go
Comment (liggitt) [inline cmd/kubelet/app/options/globalflags.go]: is it our intent to keep adding global flags like this? I expected us to manage flags in a better way going forward.

Comment (vincepri) [reply cmd/kubelet/app/options/globalflags.go]: Honestly I'm not sure. I updated the code where the codebase is referenced, I can take this one out I think, but the other change in components-base is required otherwise some tests fail.

Comment (liggitt) [reply cmd/kubelet/app/options/globalflags.go]: To be clear, the issue is with the change made in klog, but this change here is a good example of the further pollution of the global flag space it is causing. Before we pull it into kubernetes, we should resolve the future direction of klog flag handling.

Comment (liggitt) [reply cmd/kubelet/app/options/globalflags.go]: cc @dims

Comment (dims) [reply cmd/kubelet/app/options/globalflags.go]: @liggitt please see https://github.com/kubernetes/klog/pull/56 and https://github.com/kubernetes/klog/issues/55 we can continue the conversation there

**Comment by BenTheElder** [comment]
poke, are we going through with this or?

**Comment by vincepri** [comment]
@BenTheElder Just rebased and pushing, waiting for CI signal

**Comment by liggitt** [comment]
hoisting from https://github.com/kubernetes/kubernetes/pull/76474#discussion_r274951790

> To be clear, the issue is with the change made in klog, but this change here is a good example of the further pollution of the global flag space it is causing. Before we pull it into kubernetes, we should resolve the future direction of klog flag handling.

/hold

will continue in klog repo

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/klog/issues/57

**Comment by vincepri** [comment]
@liggitt @dims I updated the PR to take advantage of `klog.InitFlags(fs)`. That said I noticed that the same file (globalflags) imports `component-base/logs` which has an Init function that calls to klog https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/logs/logs.go#L35-L38, while this may fix the immediate issue with `kubelet`, it might be worth opening a different issue tracking the component-base global init.

File: staging/src/k8s.io/component-base/logs/logs.go
Comment (vincepri) [inline staging/src/k8s.io/component-base/logs/logs.go]: The above has been removed now that we're using `klog.InitFlags` elsewhere and `logtostderr` is already defaulting to true.

File: staging/src/k8s.io/component-base/cli/globalflag/globalflags_test.go
Comment (vincepri) [inline staging/src/k8s.io/component-base/cli/globalflag/globalflags_test.go]: This test seemed redundant with the one below, it was removed given that it's testing a behavior that's not true anymore given that we're not registering klog flags globally.

File: staging/src/k8s.io/component-base/logs/logs.go
Comment (liggitt) [inline staging/src/k8s.io/component-base/logs/logs.go]: drop the TODO

File: staging/src/k8s.io/component-base/logs/logs.go
Comment (liggitt) [reply staging/src/k8s.io/component-base/logs/logs.go]: is the overall init() block still required? that is still registering klog flags into the global command line, right?

**Comment by vincepri** [comment]
@dims @liggitt All tests passed 🎉, ptal

**Comment by mtaufen** [comment]
FYI it looks like this may have broken Kubelet's log to file behavior: https://github.com/kubernetes/kubernetes/issues/77416

@yujuhong mentioned it now writes to the log file up to the point where it tries to load the Kubelet config file, then starts logging to stderr afterwards. This makes me wonder if there is some nefarious interaction between how klog registers its flags and the way we currently enforce flag precedence.

**Comment by mtaufen** [comment]
@liggitt @vincepri 

I think the issue is that this PR changes the global FlagSet construction logic to call klog.InitFlags every time a new FlagSet is constructed. This has the side effect of resetting the values of all klog flags to defaults. The old logic constructed a FlagSet that referenced the global flags without changing their values.  

Constructing the global flag set should not have such a side-effect, because the Kubelet currently needs to re-parse the full command line to enforce precedence while explicitly _not_ calling Set for global flag values a second time, or else accumulator flags could end up with duplicate entries during the re-parse.

**Comment by dims** [comment]
@mtaufen also see https://github.com/kubernetes/kubernetes/issues/77305 could that be the same problem you are describing above?

File: staging/src/k8s.io/component-base/cli/globalflag/globalflags.go
Comment (mtaufen) [inline staging/src/k8s.io/component-base/cli/globalflag/globalflags.go]: This has the side effect of resetting the values in klog to defaults every time a new flagset is constructed.

**Comment by mtaufen** [comment]
@dims I believe the problem I described only happens in the Kubelet; I don't think any other components have copied the flag precedence approach in the Kubelet yet (hence the legacyflag work). That said, if other components wanted to do something similar, a similar call to InitFlags in the globalflag package in component-base will cause the same issue.

I think we should be able to fix by just reverting the Kubelet code to register each klog flag individually, and avoid the side-effect.

**Comment by mtaufen** [comment]
Alternatively, we could revert this PR, then fix klog so that InitFlags itself is free of global side-effects.

**Comment by mtaufen** [comment]
Former fix: https://github.com/kubernetes/kubernetes/pull/77562

**Comment by vincepri** [comment]
I'll defer to @liggitt and @dims on path forward, happy to put in some time for a fix if needed

**Comment by dims** [comment]
@mtaufen thanks!

@vincepri looks like @mtaufen has filed the fix for "reverting the Kubelet code", can you please do the other one? " register each klog flag individually " Thanks!

**Comment by mtaufen** [comment]
@vincepri FYI reverting the Kubelet code == registering each klog flag individually. The other one was "fix klog so that InitFlags itself is free of global side-effects."

**Pull request merged**
